### PR TITLE
Better handling of `disconnect!`/`close`.

### LINKED ISSUE #257: Occasional failures to reconnect
I have at least once instance where https://github.com/dblock/slack-sup disconnected a team and never reconnected it (nothing in logs). This is with all the fixes in #208 with 0.14.1. Restarting the bot helps obviously.

### PR DESCRIPTION
**Pull request by ioquatix**
Fixes #257

### CODE REVIEW

**Comment by dblock** [comment]
Ok, I'll take this. Care to clean it up? CHANGELOG and `rubocop -a ; rubocop --auto-gen-config`?

**Comment by ioquatix** [comment]
I am taking kids to the park so can you take it from here? Otherwise I can look at it later.

**Comment by dblock** [comment]
👪 > 👨‍💻 👩‍💻

File: lib/slack/real_time/concurrency/async.rb [hunk 1]
```diff
@@ -43,11 +43,6 @@ def restart_async(client, new_url)
             end
           end
 
-          def disconnect!
-            super
-            @reactor.cancel
```
Comment (dblock) [inline lib/slack/real_time/concurrency/async.rb]: @ioquatix Do we not need this anymore? Are we leaving the reactor running when we explicitly call `disconnect!` now?

**Comment by dblock** [comment]
I think we do, seeing this otherwise:

```
I, [2019-04-06T15:53:21.696868 #251]  INFO -- : Successfully connected team  ...
  1m2s    error: <Async::Task:0x1f3a650 connected to #<Addrinfo: 18.204.248.223:443 TCP (cerberus-xxxx.lb.slack-msgs.com)> failed> [pid=251]
               |   Async::Wrapper::Cancelled: The operation has been cancelled!
               |   → /app/vendor/bundle/ruby/2.6.0/gems/async-1.17.1/lib/async/task.rb:51 in `yield'
               |     /app/vendor/bundle/ruby/2.6.0/gems/async-1.17.1/lib/async/wrapper.rb:211 in `wait_for'
               |     /app/vendor/bundle/ruby/2.6.0/gems/async-1.17.1/lib/async/wrapper.rb:102 in `wait_readable'
               |     /app/vendor/bundle/ruby/2.6.0/gems/async-io-1.20.0/lib/async/io/generic.rb:145 in `async_send'
               |     /app/vendor/bundle/ruby/2.6.0/gems/async-io-1.20.0/lib/async/io/generic.rb:47 in `block in wrap_blocking_method'
               |     /app/vendor/bundle/ruby/2.6.0/gems/async-websocket-0.8.0/lib/async/websocket/connection.rb:60 in `next_event'
               |     /app/vendor/bundle/ruby/2.6.0/bundler/gems/slack-ruby-client-9f0860949c4f/lib/slack/real_time/concurrency/async.rb:68 in `run_loop'
               |     /app/vendor/bundle/ruby/2.6.0/bundler/gems/slack-ruby-client-9f0860949c4f/lib/slack/real_time/concurrency/async.rb:52 in `connect!'
               |     /app/vendor/bundle/ruby/2.6.0/bundler/gems/slack-ruby-client-9f0860949c4f/lib/slack/real_time/client.rb:83 in `run_loop'
               |     /app/vendor/bundle/ruby/2.6.0/bundler/gems/slack-ruby-client-9f0860949c4f/lib/slack/real_time/concurrency/async.rb:30 in `block (3 levels) in start_async'
               |     /app/vendor/bundle/ruby/2.6.0/gems/async-1.17.1/lib/async/task.rb:204 in `block in make_fiber'
```

**Comment by dblock** [comment]
Hm. That didn't do it. I need to look at how that `Async::Wrapper::Cancelled` error occurs. If you know off hand I can try to fix it.

File: lib/slack/real_time/concurrency/async.rb (hunk 1 above)
Comment (ioquatix) [reply lib/slack/real_time/concurrency/async.rb]: Let me check...

**Comment by ioquatix** [comment]
Can you add a spec that fails in the way you shown above? Otherwise it's hard to know exactly what is going on.

**Comment by dblock** [comment]
Happens on every server-side disconnect in a production bot, I tried to write one but didn't manage to.

**Comment by dblock** [comment]
Note that spec/integration/integration_spec.rb is failing with these changes, notably :112 with `fiber called across threads`.

File: lib/slack/real_time/concurrency/async.rb [hunk 2]
```diff
@@ -15,22 +15,22 @@ class Socket < Slack::RealTime::Socket
...
             Thread.new do
               @reactor.run do |task|
                 if client.run_ping?
                   task.async do |subtask|
-                    subtask.annotate "client keep-alive"
-                    
-                    while true
+                    subtask.annotate 'client keep-alive'
+
+                    loop do
```
[the remark this answers is not in the document]
Comment (dblock) [reply lib/slack/real_time/concurrency/async.rb]: That was the effect of `rubocop -a`

[the remark this answers is not in the document]
Comment (ioquatix) [reply lib/slack/real_time/concurrency/async.rb]: I know and it's objectively worse.

[the remark this answers is not in the document]
Comment (dblock) [reply lib/slack/real_time/concurrency/async.rb]:
```
Style/InfiniteLoop:
   Enabled: false
```

should fix it in rubocop.yml

**Comment by ioquatix** [comment]
Okay, I'm taking a look at the integration specs.

**Comment by ioquatix** [comment]
@dblock there is something weird going on. I'm just doing my best to analyse how things working, not trying to blame anyone here, so let's just go through it:

There are several ways a connection can be closed:

1. The driver receives a close event from the server. This should gracefully close the connection and possibly cause a restart.
2. The socket stops responding or is otherwise closed.
3. The client wants to send a close event and gracefully terminate the connection.
4. The client wants to immediately terminate the connection (i.e. call `close`).

The problem is I don't know how this is supposed to work in all 4 cases, because the flow control is a bit... odd?

From my POV, the logic is independent of the concurrency model. What I mean by that, is there should be a working implementation between `Client`, `Socket`, and `Driver`.

Let me give you an example of one case that is confusing me.

In the integration specs, `stop!` is invoked to gracefully shut down the connection (case 3 above). The way I think this should work is:

```ruby
driver.close # Send shutdown event
socket.close # Close socket file descriptor
```

Now, around that, what events do you think should be triggered?

`driver.emit(:close)` won't be triggered because we didn't *receive* a close frame, we only sent one. In  theory, it might be possible to receive a close frame between `driver.close` and `socket.close`, but let's ignore that. So, if `driver` never emits close, then `client.on(:close)` is also never triggered, and then because of that, the server closed item is never added to the integration spec `@queue`.

So, I tried to think about in what situations we want to invoke the `client.callback(:close)`, e.g. the cases 1-4 above. I get the feeling that `client.close` should be triggered in any situation where the socket is closed, not the driver. But the way it's set up right now, if you try to do that, the callback actually invokes `close` on itself. That causes infinite recursion.

Basically, `client.close` eventually invokes `socket.close`, which eventually invokes `client.close`. The only way to break this weird cycle is to nullify something... but that just seems like a hack around the broken underlying assumptions being made.

I believe we *should* still invoke `client.callback(:close)` but it should not itself call `close`. I'll try this.

**Comment by ioquatix** [comment]
Okay, so that issue was resolved by breaking the recursion.

The next problem is calling `client.stop!`. IMHO, you can only call this method from the same thread that is handling the connection. Trying to call this between threads is problematic.

**Comment by ioquatix** [comment]
There is a race condition in `Client#run_loop`.

```ruby
        @socket.connect! do |driver|
          driver.on :open do |event|
            logger.debug("#{self.class}##{__method__}") { event.class.name }
            open(event)
            callback(event, :open)
          end

          driver.on :message do |event|
            logger.debug("#{self.class}##{__method__}") { "#{event.class}, #{event.data}" }
            dispatch(event)
          end

          driver.on :close do |event|
            logger.debug("#{self.class}##{__method__}") { event.class.name }
            callback(event, :close)
            close(event)
            callback(event, :closed)
          end

          # This must be called last to ensure any events are registered before invoking user code.
          @callback.call(driver) if @callback
        end
```

We shouldn't invoke the `@callback` until all events are registered, as those events are doing some pretty critical state management (mostly `:close`).

**Comment by ioquatix** [comment]
I'm trying to understand how `disconnect!` works. In celluloid, it stops the ping timer. But is that right? Is there any normal flow control that could happen during the execution of `socket.close` which could hit this, and disable the ping timer, even though the connection still continues to run (i.e. client calls `connect!` after calling `disconnect!`.

Honestly, this state handling is super tricky.

**Comment by ioquatix** [comment]
I believe this solves all the problems and more robustly handles disconnects, but you'll probably want to test it. I didn't test event machine or celluloid implementations. Honestly, I think 99% of the logic should be implemented in the `Socket` base class, and we should just be implementing the resource management and concurrency model in the sub-class. So I tried to improve that a bit.

**Comment by dblock** [comment]
First, huge thank you @ioquatix. You're clearly following this code a lot better than I am, took me a minute to wrap my head around the changes ;)

I deployed this on a bot and on a server-side disconnect I am still seeing the same stack:

```
  1m5s    error: <Async::Task:0x2aa6d90 connected to #<Addrinfo: 18.204.248.223:443 TCP (cerberus-xxxx.lb.slack-msgs.com)> failed> [pid=250]
               |   Async::Wrapper::Cancelled: The operation has been cancelled!
               |   → /app/vendor/bundle/ruby/2.6.0/gems/async-1.17.1/lib/async/task.rb:51 in `yield'
               |     /app/vendor/bundle/ruby/2.6.0/gems/async-1.17.1/lib/async/wrapper.rb:211 in `wait_for'
               |     /app/vendor/bundle/ruby/2.6.0/gems/async-1.17.1/lib/async/wrapper.rb:102 in `wait_readable'
               |     /app/vendor/bundle/ruby/2.6.0/gems/async-io-1.20.0/lib/async/io/generic.rb:145 in `async_send'
               |     /app/vendor/bundle/ruby/2.6.0/gems/async-io-1.20.0/lib/async/io/generic.rb:47 in `block in wrap_blocking_method'
               |     /app/vendor/bundle/ruby/2.6.0/gems/async-websocket-0.8.0/lib/async/websocket/connection.rb:60 in `next_event'
               |     /app/vendor/bundle/ruby/2.6.0/bundler/gems/slack-ruby-client-da457b9b8a60/lib/slack/real_time/concurrency/async.rb:95 in `run_loop'
               |     /app/vendor/bundle/ruby/2.6.0/bundler/gems/slack-ruby-client-da457b9b8a60/lib/slack/real_time/concurrency/async.rb:71 in `connect!'
               |     /app/vendor/bundle/ruby/2.6.0/bundler/gems/slack-ruby-client-da457b9b8a60/lib/slack/real_time/client.rb:83 in `run_loop'
               |     /app/vendor/bundle/ruby/2.6.0/bundler/gems/slack-ruby-client-da457b9b8a60/lib/slack/real_time/concurrency/async.rb:48 in `block (2 levels) in start_reactor'
               |     /app/vendor/bundle/ruby/2.6.0/gems/async-1.17.1/lib/async/task.rb:204 in `block in make_fiber'
```

The reconnect happens correctly afterwards. 

This seems to make sense if I understand it correctly, as the client is waiting for a message and gets canceled. Should that just be rescued and ignored?

**Comment by ioquatix** [comment]
> First, huge thank you @ioquatix. You're clearly following this code a lot better than I am, took me a minute to wrap my head around the changes ;)

it took almost a day to create this PR, so if it only took you a minute you have one up on me :p

> Should that just be rescued and ignored? How did we not have this problem before?

I don't think you were handling the cancellation properly before.

You should probably not change it because it's actually telling you something about the failure - and this *is* an error condition that's being recovered, and so logging the exception is probably the right thing to do.

The reason why this happens is because one task is waiting for a socket to become readable, and another task cancels it, e.g.

```ruby
task = Async do
	socket.read # Will be cancelled
end

task.stop
```

What this is telling you is that the stream is still connected, but the remote end is no longer communicating.

I would say, if I had to take a guess, that somewhere upstream there is a proxy that is keeping the connection alive even thought the remote server has disconnected/timed out.

**Comment by dblock** [comment]
I see. I'll probably want to truncate logs here with a rescue/raise. 

... still need to fix specs for Celluloid/EM ... I'll get to it this week unless you beat me to it

**Comment by ioquatix** [comment]
There is zero chance I will be able to spend any more time on this for at least a month lol, sorry. Happy to answer questions.

**Comment by dblock** [comment]
Fixed concurrency specs in https://github.com/slack-ruby/slack-ruby-client/pull/262/commits/3a9df0112c4526938189ae1cdbe6217af4178a24 and reduced logger noise (removed stack) in https://github.com/slack-ruby/slack-ruby-client/pull/262/commits/3e5b22247a80b02e6b6bdd735e687fe1ae0738ca, seems to behave well in production for now. Will let it cook for a bit and merge.

```
W, [2019-04-08T11:17:32.919058 #252]  WARN -- <Async::Task:0x1e7e3ec connected to #<Addrinfo: 18.204.248.223:443 TCP (cerberus-xxxx.lb.slack-msgs.com)> running>: The operation has been cancelled!
```
